rskj-core/src/main/java/org/ethereum/rpc/LogFilter.java

```diff
@@ -231,47 +232,75 @@ private static void processSingleBlockByHash(String blockHash, Blockchain blockc
+            boolean isConfirmedBlock = block.getNumber() <= bestBlockNumber - blocksBloomStore.getNoConfirmations();
 
-                        if (!filter.addressesTopicsFilter.matchBloom(blocksBloom.getBloom())) {
-                            blockNum = blocksBloomStore.lastNumberInRange(blockNum);
-                            continue;
-                        }
-                    }
+            BlocksBloom blocksBloom = isConfirmedBlock ? blocksBloomStore.getBlocksBloomByNumber(block.getNumber()) : null;
+            boolean hasBloom = blocksBloom != null;
 
-                    auxiliaryBlocksBloom = new BlocksBloom();
-                }
+            if (hasBloom && !bloomMatches(blocksBloom, filter)) {
+                block = getBlockToSkipToOrNullIfDone(block, fromBlock.getNumber(), blocksBloomStore, blockchain);
+                continue;
+            }
 
-                Block block = blockchain.getBlockByNumber(blockNum);
+            boolean needsBloom = isConfirmedBlock && !hasBloom;
+            if (needsBloom) {
+                bloomAccumulator = addBlockBloom(block, bloomAccumulator, blocksBloomStore);
+            }
 
-                if (auxiliaryBlocksBloom != null) {
-                    auxiliaryBlocksBloom.addBlockBloom(blockNum, new Bloom(block.getLogBloom()));
-                }
+            // reverseTxOrder=true to process txs in reverse order for later complete list reverse
+            filter.onBlock(block, true);
+            block = blockchain.getBlockByHash(block.getParentHash().getBytes());
+        }
 
-                if (auxiliaryBlocksBloom != null && blocksBloomStore.lastNumberInRange(blockNum) == blockNum) {
-                    blocksBloomStore.addBlocksBloom(auxiliaryBlocksBloom);
-                }
+        // sort in a from-to fashion after looping in reverse order
+        filter.reverseEvents();
+    }
 
-                filter.onBlock(block);
-            }
-            else {
-                filter.onBlock(blockchain.getBlockByNumber(blockNum));
-            }
+    private static boolean bloomMatches(BlocksBloom blocksBloom, LogFilter filter) {
+        return filter.addressesTopicsFilter.matchBloom(blocksBloom.getBloom());
+    }
+
+    @Nullable
+    private static BlocksBloom addBlockBloom(Block block, BlocksBloom bloomAccumulator, BlocksBloomStore blocksBloomStore) {
+        // reset bloomAccumulator on every confirmed lastInRange block to start a new bloom
+        if (isLastInRange(block, blocksBloomStore)) {
+            bloomAccumulator = BlocksBloom.createEmptyWithBackwardsAddition();
+        }
+
+        // start accumulating blocks the first time a lastInRange block is reached to keep only complete blooms
+        if (bloomAccumulator == null) {
+            return null;
+        }
+
+        bloomAccumulator.addBlockBloom(block.getNumber(), new Bloom(block.getLogBloom()));
+
+        boolean firstInRange = blocksBloomStore.firstNumberInRange(block.getNumber()) == block.getNumber();

```

Comment:
is there a reason to create isLastInRange and not isFirstInRange?

Reply:
Addressed.